paddle/framework/tensor.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <typeinfo>
+#include "paddle/framework/ddim.h"
+#include "paddle/platform/assert.h"
+#include "paddle/platform/place.h"
+
+namespace paddle {
+namespace framework {
+
+class Tensor {
+  using paddle::platform::Place;
+  using paddle::platform::get_place;
+
+ public:
+  explicit Tensor(DDim dims) : dims_(dims), place_(get_place()) {}
+  explicit Tensor(DDim dims, Place place) : dims_(dims), place_(place) {}
+
+  Tensor& operator=(const Tensor& src) = delete;
+
+  template <typename T>
+  const T* Data() const {
+    PADDLE_ASSERT(holder_ != nullptr);
+    PADDLE_ASSERT(holder_->Place() == place_);
+    PADDLE_ASSERT(holder_->Size() >= product(dims_) * sizeof(T));
+    return static_cast<const T*>(holder->Ptr());
+  }
+
+  template <typename T>
+  bool NeedReset() const {
+    return (holder_ == nullptr || holder_->Place() != place_ ||
+            holder_->Size() < product(dims_) * sizeof(T));
+  }
+
+  // must be POD types
+  template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+  T* MutableData() {
+    if (NeedReset<T>()) {
+      holder_.reset(new PlaceholderImpl(place_, product(dims_) * sizeof(T)));
+    }
+    return static_cast<T*>(holder_->Ptr());
+  }
+
+  template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+  T* MutableData(const DDim& dims) {
+    dims_ = dims;
+    return MutableData<T>();
+  }
+
+  template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+  T* MutableData(const DDim& dims, const Place& place) {
+    dims_ = dims;
+    place_ = place;
+    return MutableData<T>();
+  }
+
+  int Rank() const;
+
+  int Numel() const;
+
+  void Resize(const DDim& dims);
+
+  void Reshape(const DDim& dims);
+
+  template <typename T>
+  void ShareData(const Tensor& src) {
+    if (src.NeedReset<T>()) {
+      // TODO: error: "Src tensor need to be reseted before calling
+      // ShareData()".
+    }
+    holder_ = src.Holder();
+    dims_ = src.Dims();
+    place_ = src.Place();
+    return;
+  }
+
+  template <typename T>
+  void CopyFrom(const Tensor& src) {
```

Comment:
Sounds like what we need is a method `Serialize` that fills in a protobuf message?

```cpp
void Tensor::Serialize(TensorProto* proto) {
  if (is_gpu_place(Place()) {
    Tensor cpu_tensor;
    cudaMemcpy(cpu_tensor.mutable_data(Size(), CPUPlace()), data(), Size());
    cpu_tensor.Serialize(proto);
  } else {
    // fill in proto
  }
}
```